src/storage/src/types/errors.rs

```diff
@@ -50,22 +50,200 @@ impl Codec for DecodeError {
         "protobuf[DecodeError]".into()
     }
 
-    fn encode<B: BufMut>(&self, buf: &mut B) {
+    fn encode<B>(&self, buf: &mut B)
+    where
+        B: BufMut,
+    {
         self.into_proto()
             .encode(buf)
-            .expect("no required fields means no initialization errors");
+            .expect("no required fields means no initialization errors")
     }
 
-    fn decode(buf: &[u8]) -> Result<Self, String> {
+    fn decode<'a>(buf: &'a [u8]) -> Result<Self, String> {
         let proto = ProtoDecodeError::decode(buf).map_err(|err| err.to_string())?;
         proto.into_rust().map_err(|err| err.to_string())
     }
 }
 
 impl Display for DecodeError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match &self.raw {
+            None => write!(f, "{}", self.inner),
+            Some(raw) => write!(f, "{} (original bytes: {:x?})", self.inner, raw),
+        }
+    }
+}
+
+#[derive(Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash)]
+pub enum DecodeErrorInner {
+    Text(String),
+}
+
+impl RustType<ProtoDecodeErrorInner> for DecodeErrorInner {
+    fn into_proto(&self) -> ProtoDecodeErrorInner {
+        use proto_decode_error_inner::Kind::*;
+        ProtoDecodeErrorInner {
+            kind: Some(match self {
+                DecodeErrorInner::Text(v) => Text(v.clone()),
+            }),
+        }
+    }
+
+    fn from_proto(proto: ProtoDecodeErrorInner) -> Result<Self, TryFromProtoError> {
+        use proto_decode_error_inner::Kind::*;
+        match proto.kind {
+            Some(Text(v)) => Ok(DecodeErrorInner::Text(v)),
+            None => Err(TryFromProtoError::missing_field("ProtoDecodeError::kind")),
+        }
+    }
+}
+
+impl Display for DecodeErrorInner {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            DecodeErrorInner::Text(e) => write!(f, "Text: {}", e),
+        }
+    }
+}
+
+/// Errors arising during envelope processing.
+#[derive(Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash)]
```

Comment:
why do we still need `Serialize` and `Deserialize`?

Reply:
Everything that's the input or output of dataflow operators needs to be serializable because Timely might need to send it across the wire to or from an operator on another machine.